### ISSUE TITLE
proposal: read -last <duration> should quantize to the specified time unit

### ISSUE THREAD

**Issue by demmer** [juttle/juttle#652]
The current semantics for the `-last` option to `read` are that `read -last <duration>` is equivalent to `read -from :now: - <duration> -to :now:`.

While simple to implement and to describe, this is not always the most intuitive behavior, since the time intervals do not align to natural boundaries. This proposal is to change the semantics so that the time period for the query aligns to the units specified in the `-last` duration. 

In other words, `read -last <N units>` would be equivalent to `-from Date.startOf(:now:, units) - :N units: -to Date.startOf(:now:, units)`.

This would mean that `read -last :day:` would result in data for the whole previous day, i.e. `-from :yesterday: -to :today:`, and not the time period starting 24 hours before the millisecond when the program was run and up to the current time, `read -last :week:` would be the whole previous week (whether it ends on Thursday or Sunday is orthogonal), and `read -last :year:` would read the whole prior calendar year, i.e. all of 2015.

In order to make this work reasonably, we would need to disallow durations that do not round to a single unit. For example, `read -last :130 minutes:` would be allowed, but `read -last :2 hours and 10 minutes:` would be rejected, even though the specified time span is the same in both expressions.

Implementation-wise this wouldn't be too challenging -- we'd need to modify moment-parser to preserve the units used in the original expression and then pass it through the generator to be stashed in the JuttleMoment Duration, then add the requisite checks and logic to the `read` proc.

**Comment by demmer** [comment]
@dmehra / @go-oleg As we discussed this am.

@dmajda @davidvgalbraith @mstemm @VladVega @bkutil @mccanne what do you guys think about this?

**Comment by mstemm** [comment]
I think having -last quantized like this is misleading. I would be confused if I did a read with -last :day: and nothing for the current day was returned.

Maybe we could have a separate option or a modifier to -last? Something like -last-whole or -last -round or something?

**Comment by dmehra** [comment]
@go-oleg proposed `-past` instead of `-last`...

**Comment by rlgomes** [comment]
:-1: completely changing the behavior of the already in use `-last` feels bad and quite difficult to retrain our brains at this point in time (also there must be quite a few tests/examples/demos to update). "Something happened in the last 24 hours" doesn't mean to me that it happened from 1 day ago at midnight till midnight of the current day. The `-last-whole` could be one solution otherwise maybe the moment-parser could be slightly modified to allow for `:1 whole/complete day ago:` ?

**Comment by dmehra** [comment]
@rlgomes and I also discussed what this would do to moment math (not something i considered earlier today), would computed expressions like `:1 day: + :12 hours:` also be disallowed in `-last X`? That seems too restrictive.

Also, having considered the impact on existing users, i'm ready to retract the proposal. For example, all of AWS cloudwatch examples would be affected. And looking at the most recent data (common case) would get clunkier, necessitating `-from :1 hour ago: -to :now:` syntax instead of the shorthand people are used to.

Still curious if supporting `-past X` with the proposed semantics would be a good compromise (i.e., `read -past :year:` means all of 2015, none of 2016, as opposed to `read -last :year:` which means 1 year back from current date). Then google-analytics adapter, which kicked off this conversation, could only support `read ga -past X` and reject `read ga -last X`.

**Comment by dmajda** [comment]
Strong :-1: on my part. I think the suggested behavior would be non-intuitive and highly confusing for users. The current behavior is also desired more often than the suggested one, so this would make the common case harder and the less common case easier.

The `-past` proposal makes sense to me, although it still has come potential for confusion.

> Then google-analytics adapter, which kicked off this conversation, could only support `read ga -past X` and reject `read ga -last X`.

What exactly triggered this? (What the GA API can’t do?)

Also, was there ever a need for this feature outside this specific use case?

**Comment by demmer** [comment]
> What exactly triggered this? (What the GA API can’t do?)

The GA API only accepts time range filters in whole day intervals. Therefore the adapter needs to restrict the `-from` and `-to` options so that they only specify times that align with the day. This means that in practice a user will almost never be able to use `-last <duration>` with the GA adapter, so we may as well forbid it altogether.

> Also, was there ever a need for this feature outside this specific use case?

During the discussion about the limitations in GA, @dmehra mentioned that she felt it to be more natural if `read x -last :day:` would return the previous full day, not the previous 24 hour timespan, hence spawning this whole thread. Given that she's retracted support for this idea and the chorus of strong opposition, I'll amend this proposal to describe another time range option (`-past / -previous / -prior` would all work in my book) rather than change the semantics for `-last`.

**Comment by davidbcook** [comment]
I agree with everything that's been said so far against this proposal; in other words, please don't change `-last`. Furthermore, the work to add a `-past` option doesn't seem worth it since if you want whole time periods, you can `reduce -every :period:` to group things into the desired period. If you do decide to build that, I'd go with Rodney's suggest of `-lastWhole` instead of the other suggestions because those seem to easily confused with what `-last` already does.
